src/box/box.cc

```diff
@@ -4296,3 +4297,264 @@ box_read_ffi_enable(void)
 	if (--box_read_ffi_disable_count == 0)
 		box_read_ffi_is_disabled = false;
 }
+
+/**
+ * Encode a sequence of values according to format string into a buffer
+ * allocated on region.
+ */
+static const char *
+mp_format_on_region(struct region *region, const char *format, ...)
```

Comment:
The same is done in `boxk()`. I suggest to use it there too. Should be easy given that it is in the same file (box.cc).

Reply:
Thanks for the review! I tried to use `mp_format_on_region()` but it's quite problematic as we can't call one variadic function from another. We can use `ap_list` instead of `...` for `mp_format_on_region()`, but we have to create a new function that creates `ap_list` for `constraint_create()` and `constraint_drop()`. Not sure if this is needed.

I added size as a new output argument to mp_format_on_region().

Reply:
Try this:
<details><summary>Diff</summary>

```Diff
--- a/src/box/box.cc
+++ b/src/box/box.cc
@@ -2312,6 +2312,41 @@ box_set_txn_isolation(void)
 
 /* }}} configuration bindings */
 
+/**
+ * Encode a sequence of values according to format string into a buffer
+ * allocated on region.
+ */
+static const char *
+mp_vformat_on_region(struct region *region, size_t *size, const char *format,
+		     va_list src)
+{
+	va_list ap;
+	va_copy(ap, src);
+	size_t buf_size = mp_vformat(NULL, 0, format, ap);
+	char *buf = (char *)region_alloc(region, buf_size);
+	va_end(ap);
+	if (buf == NULL) {
+		diag_set(OutOfMemory, buf_size, "region_alloc", "buf");
+		return buf;
+	}
+	va_copy(ap, src);
+	mp_vformat(buf, buf_size, format, ap);
+	va_end(ap);
+	*size = buf_size;
+	return buf;
+}
+
+static const char *
+mp_format_on_region(struct region *region, size_t *size, const char *format,
+		    ...)
+{
+	va_list ap;
+	va_start(ap, format);
+	const char *res = mp_vformat_on_region(region, size, format, ap);
+	va_end(ap);
+	return res;
+}
+
 /**
  * Execute a request against a given space id with
  * a variable-argument tuple described in format.
@@ -2335,17 +2370,13 @@ boxk(int type, uint32_t space_id, const char *format, ...)
 	request.type = type;
 	request.space_id = space_id;
 	va_start(ap, format);
-	size_t buf_size = mp_vformat(NULL, 0, format, ap);
-	char *buf = (char *)region_alloc(&fiber()->gc, buf_size);
+	struct region *region = &fiber()->gc;
+	size_t size = 0;
+	const char *data = mp_vformat_on_region(region, &size, format, ap);
 	va_end(ap);
-	if (buf == NULL)
+	if (data == NULL)
 		return -1;
-	va_start(ap, format);
-	if (mp_vformat(buf, buf_size, format, ap) != buf_size)
-		assert(0);
-	va_end(ap);
-	const char *data = buf;
-	const char *data_end = buf + buf_size;
+	const char *data_end = data + size;
 	switch (type) {
 	case IPROTO_INSERT:
 	case IPROTO_REPLACE:
@@ -4298,30 +4329,6 @@ box_read_ffi_enable(void)
 		box_read_ffi_is_disabled = false;
 }
 
-/**
- * Encode a sequence of values according to format string into a buffer
- * allocated on region.
- */
-static const char *
-mp_format_on_region(struct region *region, uint32_t *size, const char *format,
-		    ...)
-{
-	va_list ap;
-	va_start(ap, format);
-	size_t buf_size = mp_vformat(NULL, 0, format, ap);
-	char *buf = (char *)region_alloc(region, buf_size);
-	va_end(ap);
-	if (buf == NULL) {
-		diag_set(OutOfMemory, buf_size, "region_alloc", "buf");
-		return buf;
-	}
-	va_start(ap, format);
-	mp_vformat(buf, buf_size, format, ap);
-	va_end(ap);
-	*size = buf_size;
-	return buf;
-}
-
 /**
  * Drop the constraint or foreign key. If fieldno is equal to UINT32_MAX then
  * the tuple constraint or tuple foreign key is dropped, otherwise the field
@@ -4378,7 +4385,7 @@ constraint_drop(uint32_t space_id, const char *name, uint32_t name_len,
 		snprintf(buf, size, "%s.%*s", path, name_len, name);
 		path = buf;
 	}
-	uint32_t ops_size;
+	size_t ops_size;
 	const char *ops = mp_format_on_region(region, &ops_size, "[[%s%s%u]]",
 					      "#", path, 1);
 	const char *end = ops + ops_size;
@@ -4485,7 +4492,7 @@ constraint_create(uint32_t space_id, const char *name, uint32_t name_len,
 	struct region *region = &fiber()->gc;
 	uint32_t used = region_used(region);
 	const char *val;
-	uint32_t val_size;
+	size_t val_size;
 	assert(type == CONSTR_FUNC || type == CONSTR_FKEY);
 	if (type == CONSTR_FUNC) {
 		val = mp_format_on_region(region, &val_size, "%u",
@@ -4502,7 +4509,7 @@ constraint_create(uint32_t space_id, const char *name, uint32_t name_len,
 	if (val == NULL)
 		return -1;
 	const char *ops;
-	uint32_t ops_size;
+	size_t ops_size;
 	if (is_empty) {
```
</details>

Reply:
Thank you for the diff! Applied.